### PR TITLE
Bug 1908347: Apply ValidatingWebhookConfiguration correctly

### PR DESCRIPTION
**Pull request by jsafrane**
The webhook should not be applied by CVO, because CVO overwrites injected CA bundle, new CA bundle gets injected and CVO overwrites in an endless (but very slow) loop.

The operator should merge the existing webhook (with CA bundle) with the version it wants to have, as it does with the other API objects. Use newly introduced `ApplyValidatingWebhookConfiguration()`.

@openshift/storage

### CODE REVIEW

**Comment by jsafrane** [comment]
Webhook runs in CI and validates requests: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-csi-snapshot-controller-operator/68/pull-ci-openshift-cluster-csi-snapshot-controller-operator-master-e2e-aws/1336394882062749696/artifacts/e2e-aws/gather-extra/pods/openshift-cluster-storage-operator_csi-snapshot-webhook-6795678f78-9zwl8_webhook.log

**Comment by jsafrane** [comment]
Note: this already includes #70

File: pkg/operator/webhookdeployment/webhook.go [hunk 1]
```diff
@@ -85,6 +104,13 @@ func (c *csiSnapshotWebhookController) sync(ctx context.Context, syncCtx factory
 		return err
 	}
 
+	webhookConfig, err := getWebhookConfig()
+	if err != nil {
+		return err
+	}
+	lastWebhookGeneration := resourcemerge.ExpectedValidatingWebhooksConfiguration(webhookConfig.Name, opStatus.Generations)
+	webhookConfig, _, err = resourceapply.ApplyValidatingWebhookConfiguration(c.kubeClient.AdmissionregistrationV1(), syncCtx.Recorder(), webhookConfig, lastWebhookGeneration)
```
Comment (bertinatto) [inline pkg/operator/webhookdeployment/webhook.go:112]: `err` isn't checked anywhere

File: pkg/operator/webhookdeployment/webhook.go [hunk 2]
```diff
@@ -149,3 +176,16 @@ func (c *csiSnapshotWebhookController) getDeployment(opSpec *operatorapi.Operato
...
 }
+
+func getWebhookConfig() (*admissionv1.ValidatingWebhookConfiguration, error) {
+	webhookBytes := generated.MustAsset(webhookAsset)
+	requiredObj, err := runtime.Decode(admissionCodecs.UniversalDecoder(admissionv1.SchemeGroupVersion), webhookBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	webhook := requiredObj.(*admissionv1.ValidatingWebhookConfiguration)
```
Comment (bertinatto) [inline pkg/operator/webhookdeployment/webhook.go:187]: in a perfect world with infinite time we'd move this to library-go's resourceread package :)

Comment (jsafrane) [reply pkg/operator/webhookdeployment/webhook.go:187]: created https://github.com/openshift/library-go/pull/972

File: pkg/operator/webhookdeployment/webhook.go (hunk 1 above)
Comment (jsafrane) [reply pkg/operator/webhookdeployment/webhook.go:112]: fixed

**Comment by gnufied** [comment]
@jsafrane while this change looks fine. Should there be a bug against CVO so it handles this correctly and does not overwrite injected CA bundle blindly?

File: assets/webhook_config.yaml [hunk 3]
```diff
@@ -14,11 +13,14 @@ webhooks:
       service:
         name: csi-snapshot-webhook
         namespace: openshift-cluster-storage-operator
-        path: "/volumesnapshot"
+        path: /volumesnapshot
     rules:
       - operations: [ "CREATE", "UPDATE" ]
         apiGroups: ["snapshot.storage.k8s.io"]
-        apiVersions: ["v1", "v1beta1"]
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/webhook_config.yaml]: yes, I rebased this PR on top of #70

**Comment by jsafrane** [comment]
> Should there be a bug against CVO so it handles this correctly and does not overwrite injected CA bundle blindly?

I don't think so... CVO and its release manifets are the ultimate source of truth for every other object deployed by it, I don't think adding exception for WebhookConfigs would be useful.

**Review by bertinatto — COMMENTED** [review]
/lgtm

**Pull request merged**